File: polyglot/runtime_template_resolver/py/src/runtime_template_resolver/coercion.py
import json
from typing import Any
from .logger import logger

log = logger.create("runtime_template_resolver", __file__)

def to_string(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, (dict, list)):
        try:
            return json.dumps(value)
        except Exception as e:
            log.warning("Failed to stringify object", error=str(e))
            return str(value)
    return str(value)

File: polyglot/runtime_template_resolver/mjs/src/coercion.ts
import { logger } from './logger.js';

const log = logger.create('runtime-template-resolver', 'coercion.ts');

export function toString(value: unknown): string {
    if (value === null || value === undefined) {
        return '';
    }
    if (typeof value === 'object') {
        try {
            return JSON.stringify(value);
        } catch (e) {
            log.warn('Failed to stringify object', { error: String(e) });
            return '[object Object]';
        }
    }
    return String(value);
}
